File: Adv_DataStructures/Trie/trie_implementation.py
# Name: Isaac Watts
# Date: 04/17/24
# Description: 
# Implementing a trie with the words apart, apple, bin, binary, bit, byte, day, days, deep, and Isaac
# Searching for a thing not in the trie and a thing that is not in the trie


class TrieNode:
    def __init__(self):
        self.children = {}

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def search(self, word):
        currentNode = self.root

        for char in word:
            if currentNode.children.get(char):
                currentNode = currentNode.children[char]
            else:
                return None
        return currentNode
    
    def insert(self, word):
        currentNode = self.root

        for char in word:
            if currentNode.children.get(char):
                currentNode = currentNode.children[char]
            else:
                newNode = TrieNode()
                currentNode.children[char] = newNode

                currentNode = newNode
        currentNode.children["*"] = None


if __name__ == '__main__':
    TrieData = Trie()
    words = ["apart", "apple", "bin", "binary", "bit", "byte", "day", "days", "deep", "Isaac"]
    for word in words:
        TrieData.insert(word.lower())
    

    Test = ["binary", "pear"]
    for t in Test:
        if TrieData.search(t.lower()) is None:
            print(f"'{t}' is not in the trie")
        else:
            print(f"'{t}' is in the trie")
    

File: Adv_DataStructures/Trie/trie-implementation.cpp
#include <iostream>
#include <vector>
#include <map>

class TrieNode {
public:
    TrieNode() {}
    std::map<char, TrieNode*> children;
};

class Trie {
public:
    Trie() {
        root = new TrieNode();
    }

    TrieNode* search(std::string word) {
        TrieNode* currentNode = root;

        for (const auto& c : word) {
            if (currentNode->children.find(c) != currentNode->children.end()) {
                currentNode = currentNode->children[c];
            }
            else {
                return nullptr;
            }
        }
        return currentNode;
    }

    void insert(std::string word) {
        TrieNode* currentNode = root;

        for (const auto& c : word) {
            if (currentNode->children.find(c) != currentNode->children.end()) {
                currentNode = currentNode->children[c];
            }
            else {
                TrieNode* newNode = new TrieNode();
                currentNode->children[c] = newNode;
                currentNode = newNode;
            }
        }
        currentNode->children['*'] = nullptr;
    }

private:
    TrieNode* root;
};

int main() {
    Trie TrieData;
    std::vector<std::string> words = {
        "apart", "apple", "bin", "binary", "bit", "byte", "day", "days", "deep", "Isaac"
    };

    for (const auto& word : words) {
        TrieData.insert(word);
    }

    std::vector<std::string> Test = {
        "apart", "apple", "bin", "binary", "bit", "byte", "day", "days", "deep", "Isaac", "hey"
    };

    for (const auto& t : Test) {
        if (TrieData.search(t) == nullptr) {
            std::cout << t << " is not in the Trie" << std::endl;
        }
        else {
            std::cout << t << " is in the Trie" << std::endl;
        }
    }
    return 0;
}
